hs_core/views/resource_rest_api.py

```diff
@@ -0,0 +1,622 @@
+from rest_framework.views import APIView
+from rest_framework.response import Response
+from rest_framework import status, generics
+from rest_framework.exceptions import *
+from hs_core import hydroshare
+from hs_core.hydroshare import file_size_limit_for_display
+from hs_core.views import utils
+import simplejson as sj
+import serializers
+from hs_core.signals import *
+import os
+
+class ResourceRetrieveUpdateDelete(APIView):
+    """
+    Retrieve, update or delete a resource
+
+    REST URL: hsapi/resource/{pk}
+
+    HTTP method: GET
+
+    HTTP status code (on success):200
+
+    :return: (on success): resource bag file if request content-type is 'application/zip' otherwise resource science metadata file
+
+    REST URL: hsapi/resource/{pk}
+
+    HTTP method: DELETE
+
+    HTTP status code (on success):200
+
+    :return: (on success): json string of the format: {'resource_id':pk}
+
+    REST URL: hsapi/resource/{pk}
+
+    HTTP method: PUT
+
+    HTTP status code (on success):200
+
+    :return: (on success): json string of the format: {'resource_id':pk}
+
+    :type   str
+
+    :param  pk: resource id
+
+    :rtype:  json string for http methods DELETE and PUT, and resource file data bytes for GET
+
+    :raises:
+
+
+    &nbsp;&nbsp;&nbsp;NotFound: return json format: {'detail': 'No resource was found for resource id':pk}
+
+    &nbsp;&nbsp;&nbsp;HTTP status code: 404
+
+    &nbsp;&nbsp;&nbsp;PermissionDenied: return json format: {'detail': 'You do not have permission to perform this action.'}
+
+    &nbsp;&nbsp;&nbsp;HTTP status code:403
+
+    &nbsp;&nbsp;&nbsp;ValidationError: return json format: {parameter-1': ['error message-1'], 'parameter-2': ['error message-2'], .. }
+
+    &nbsp;&nbsp;&nbsp;HTTP status code:400
+    """
+
+    @property
+    def allowed_methods(self):
+        return ['GET', 'PUT', 'DELETE']
+
+    def get(self, request, pk):
+        # check the user is authorized for this action and if the resource exists for the given resource id (pk)
+        utils.authorize(request, pk, view=True)
+        resource_bag = hydroshare.get_resource(pk)
+
+        if request.content_type == 'application/zip':
+            # redirects to django_irods/views.download function
+            return HttpResponseRedirect(resource_bag.bag.url)
+        else:
+            return redirect('science_metadata_get_update', pk=pk)
+
+    def put(self, request, pk):
+        # check the user is authorized for this action and if the resource exists for the given resource id (pk)
+        utils.authorize(request, pk, edit=True)
+
+        resource_update_request_validator = serializers.ResourceUpdateRequestValidator(data=request.data)
+        if not resource_update_request_validator.is_valid():
+            raise ValidationError(detail=resource_update_request_validator.errors)
+
+        # TODO: update resource - not sure what exactly we need to update
+        raise NotImplementedError()
```

Comment:
This reminds me of the new bagit workflow to be worked on after the next release. In the new bagit workflow, bag creation will be done on iRODS side and we store each individual file in the bag for iRODS rule to create a zipped bag. This new bagit workflow may affect this specific web service API, i.e., rather than passing in a bag to create a resource, separate sequence of APIs might be needed to allow users to create a resource initially, then add files into them, or something like that. Just a thought.